test/core/event_engine/test_suite/tests/dns_test.cc

```diff
@@ -25,7 +68,747 @@ void InitDNSTests() {}
+    // environment.
+    std::string kPythonWrapper = "tools/distrib/python_wrapper";
+    std::string kDNSServerPath = "test/cpp/naming/utils/dns_server";
+    std::string kDNSResolverPath = "test/cpp/naming/utils/dns_resolver";
+    std::string kTCPConnectPath = "test/cpp/naming/utils/tcp_connect";
+    // HACK: Hyrum's law.
+    if (!grpc_core::GetEnv("TEST_SRCDIR").has_value()) {
+      // Invoke the .sh and .py scripts directly where they are in source code.
+      kPythonWrapper += ".sh";
+      kDNSServerPath += ".py";
+      kDNSResolverPath += ".py";
+      kTCPConnectPath += ".py";
+    }
+    // 1. launch dns_server
+    int port = grpc_pick_unused_port_or_die();
+    // <path to python wrapper> <path to dns_server.py> -p <port> -r <path to
+    // records config>
+    _dns_server.server_process = new grpc::SubProcess(
+        {kPythonWrapper, kDNSServerPath, "-p", std::to_string(port), "-r",
+         kDNSTestRecordGroupsYamlPath});
+    _dns_server.port = port;
+
+    // 2. wait until dns_server is up (health check)
+    bool health_check_succeed = false;
+    for (int i = 0; i < 10; i++) {
+      // 2.1 tcp connect succeeds
+      // <path to python wrapper> <path to tcp_connect.py> -s <hostname> -p
+      // <port>
+      grpc::SubProcess tcp_connect({kPythonWrapper, kTCPConnectPath, "-s",
+                                    "localhost", "-p", std::to_string(port)});
+      int status = tcp_connect.Join();
+      // TODO(yijiem): make this portable for Windows
+      if (WIFEXITED(status) && WEXITSTATUS(status) == 0) {
+        // 2.2 make an A-record query to dns_server
+        // <path to python wrapper> <path to dns_resolver.py> -s <hostname> -p
+        // <port> -n <domain name to query>
+        std::string command = absl::StrJoin<absl::string_view>(
+            {kPythonWrapper, kDNSResolverPath, "-s", "127.0.0.1", "-p",
+             std::to_string(port), "-n", kHealthCheckRecordName},
+            " ");
+        // TODO(yijiem): make this portable for Windows
+        FILE* f = popen(command.c_str(), "r");
+        GPR_ASSERT(f != nullptr);
+        char buf[128] = {};
+        size_t res =
+            fread(buf, sizeof(buf[0]), sizeof(buf) / sizeof(buf[0]) - 1, f);
+        if (pclose(f) == 0 && res > 0) {
+          absl::string_view sv(buf);
+          if (sv.find("123.123.123.123") != sv.npos) {
+            // finally
+            gpr_log(
+                GPR_INFO,
+                "DNS server is up! Successfully reached it over UDP and TCP.");
+            health_check_succeed = true;
+            break;
+          }
+        }
+      }
+      absl::SleepFor(absl::Seconds(1));
+    }
+    ASSERT_TRUE(health_check_succeed);
+  }
+
+  static void TearDownTestSuite() {
+    _dns_server.server_process->Interrupt();
+    _dns_server.server_process->Join();
+    delete _dns_server.server_process;
+  }
+
+ protected:
+  // Verifies that an ExecCtx exists on the thread from which the destructor is
```

Comment:
As per my comment elsewhere, I don't think this is actually the invariant that we want.

I think part of the confusion here may be that we are calling `EventEngine::Run()` from inside of the `EventEngine` impl.  But I don't think our rule is that any caller of `EE::Run()` needs to create an `ExecCtx` in the callback; in particular, I do *not* think that's appropriate if `EE::Run()` is called from within the EE impl.  I think the rule is actually that any caller of *any* EE API *from outside of the EE impl* needs to create an `ExecCtx` in its callback.

In other words, the internal callers to `EE::Run()` should *not* create an `ExecCtx` in the callback, but the external callers to the methods of `DNSResolver` (including the ones in this test) *should* create an `ExecCtx` in the callback.

Reply:
Done. I removed all `ExecCtx`s in EE's `grpc_ares_wrapper` and `PosixDNSResolver` and the verifier in this test. I also made changes in `grpc_ares_wrapper` so that if we are on an EventEngine thread, we will invoke the `on_resolve` callback inline instead of calling `EE::Run()` again.